### PR TITLE
Custom route formatters

### PR DESCRIPTION
**Pull request by phendrick**
### Summary
To allow for custom route renderers (to add colour, emoji, or additional route filtering) via the `rails routes` command, this PR adds formatter registration to the ConsoleFormatters module. 

The `Sheet` and `Expanded` formatters are now `SheetFormatter` and `ExpandedFormatter` and are both registered formatters. 

### Other Information

I've written a small demo gem which you can see [here](https://github.com/phendrick/demo_route_formatter)
which registers an emoji formatter and renders this output when running `rails routes --formatter emoji`: 

Image of Emoji routes formatter output

Generating the output of the `routes` command via a custom formatter means we can add additional flags, like `--compact` which strips out certain routes like those that belong to action_mailbox, turbo etc)

### CODE REVIEW

**Comment by skipkayhil** [comment]
Very related: #45241, it may be worth trying to work together on this one

**Comment by phendrick** [comment]
> Very related: #45241, it may be worth trying to work together on this one

Thanks @skipkayhil I'll follow up on that PR 👍

File: actionpack/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,16 @@
+*   Adds route formatter registration to `ActionDispatch::Routing::ConsoleFormatter`
+
+    To allow for easier cusomisation of the output from `rails routes` this allows for
```
Comment (rafaelfranca) [inline actionpack/CHANGELOG.md:3]:
```suggestion
    To allow for easier customization of the output from `rails routes` this allows for 
```

File: actionpack/lib/action_dispatch/routing/inspector.rb [hunk 2]
```diff
@@ -134,111 +126,134 @@ def collect_engine_routes(route)
...
     module ConsoleFormatter
+      cattr_accessor :registered_formatters, default: {}
+
+      def self.register_formatter(formatter, name = nil)
+        formatter_name = name || formatter.name.split(/::/).last
+        registered_formatters[formatter_name] = formatter if valid_formatter?(formatter)
+      end
+
+      def self.valid_formatter?(formatter)
+        raise ArgumentError, "#{formatter} doesn't implement #section" unless formatter.instance_methods.include?(:draw)
```
Comment (rafaelfranca) [inline actionpack/lib/action_dispatch/routing/inspector.rb:137]:
```suggestion
        raise ArgumentError, "#{formatter} doesn't implement #draw" unless formatter.instance_methods.include?(:draw)
```

File: actionpack/lib/action_dispatch/routing/inspector.rb [hunk 3]
```diff
@@ -134,111 +126,134 @@ def collect_engine_routes(route)
...
-            name_width, verb_width, path_width = widths(routes).zip(header_lengths).map(&:max)
+    ConsoleFormatter.register_formatter(SheetFormatter, "default")
 
-            routes.map do |r|
-              "#{r[:name].rjust(name_width)} #{r[:verb].ljust(verb_width)} #{r[:path].ljust(path_width)} #{r[:reqs]}"
-            end
-          end
+    class ExpandedFormatter < ConsoleFormatter::Base
+      def initialize(width: IO.console_size[1])
+        @width = width
```
Comment (rafaelfranca) [inline actionpack/lib/action_dispatch/routing/inspector.rb:218]: Doesn't Base already set the width in the initialize?

**Pull request closed without merging**